addons/udes_stock/tests/test_stock_reservation.py

```diff
@@ -95,3 +96,222 @@ def test_logs_debug_message_if_more_one_picking_processed(self):
+# are: sufficent stock, no stock, some lines are not reservable and some lines
+# are partially available.
+
+# To test this, we have a matrix of configuration options, initial stock levels
+# and expected reservations. The ReservationMixin class has tests for each of
+# the stock level conditions; the TestCase classes provide the configuration
+# for each test method.
+
+# Test configuration, stock levels and expected outputs, keyed on:
+# Pickings are in batches, PickingType.u_reserve_batches, PickingType.u_handle_partials.
+# Values order: all available, none available, partially available, line partially available
+# In-values order:
+# quantity in stock, expected reservation
+# each picking has two identical lines (2 x 10)
+# each batch has two identical pickings
+RESERVATION_MATRIX = {
+    (True, True, True): [(40, 40), (0, 0), (30, 30), (35, 35)],
+    (True, True, False): [(40, 40), (0, 0), (0, 0), (0, 0)],
+    (True, False, True): [(40, 40), (0, 0), (30, 30), (35, 35)],
+    (True, False, False): [(40, 40), (0, 0), (30, 20), (35, 20)],
+    (False, True, True): [(40, 40), (0, 0), (40, 40), (40, 40)],
+    (False, True, False): [(40, 40), (0, 0), (0, 0), (0, 0)],
+    (False, False, True): [(40, 40), (0, 0), (30, 30), (35, 35)],
+    (False, False, False): [(40, 40), (0, 0), (0, 0), (0, 0)],
+}
+
+
+class ReservationMixin(object):
+    def test_reserves_stock_when_sufficient_stock_is_available(self):
+        """The system will reserve all demand if stock is available."""
+        initial_qty, expected = self.config[0]
+        self.setup(initial_qty)
+
+        with mock.patch.object(self.pick.env.cr, "commit", return_value=None):
+            self.pick.reserve_stock()
+
+        quant = self.Quant.search([("reserved_quantity", ">", 0)])
+        self.assertEqual(quant.reserved_quantity, expected)
+
+    def test_does_not_reserve_if_no_available_stock(self):
+        """The system will not reserve stock if none is available."""
+        initial_qty, expected = self.config[1]
+        self.setup(initial_qty)
+
+        with mock.patch.object(self.pick.env.cr, "commit", return_value=None):
+            self.pick.reserve_stock()
+
+        quant = self.Quant.search([("reserved_quantity", ">", 0)])
+        self.assertEqual(quant.reserved_quantity, expected)
+
+    def test_handles_partially_available_pickings(self):
+        """The system will correctly handle the case when not all lines can be reserved."""
+        initial_qty, expected = self.config[2]
+        self.setup(initial_qty)
+
+        with mock.patch.object(self.pick.env.cr, "commit", return_value=None):
+            self.pick.reserve_stock()
+
+        quant = self.Quant.search([("reserved_quantity", ">", 0)])
+        self.assertEqual(quant.reserved_quantity, expected)
+
+    def test_handles_partially_available_lines(self):
+        """The system will correctly handle the case when a line is partially available."""
+        initial_qty, expected = self.config[2]
```

Comment:
Config mostly is getting the third element while when defined it seems has 4 elements. Do we need the fourth one or i am missing smth

Reply:
Each value in the `RESERVATION_MATRIX` dict consists for four tuples, that correspond to the states of 

- enough stock to reserve
- no stock to reserve
- enough stock for all lines except one
- enough stock for ll lines except one that we can partially reserve

Each of the tests in `ReservationMixin` tests one of these states, and fetches the corresponding element element from the value list.

Reply:
Ah I see - that last test should be `[3]` (thanks @ws-mps )

Reply:
Fixed by 0c72542e2, but I've since reworked that test case so that it hast just one method which loops over the config values.